mdnsbridge/mdnsbridgeclient.py

```diff
@@ -53,22 +61,35 @@ def getHref(self, srv_type, priority=None, api_ver=None, api_proto=None):
         if srv_type not in self.services:
             self.services[srv_type] = []
 
+        # Flush the cached list of services
+        if flush:
+            self.updateServices(srv_type)
+
         # Check if there are any of that type of service, if not do a request
-        no_results = True
-        for service in self.services[srv_type]:
-            if api_ver is not None and api_ver not in service["versions"]:
-                continue
-            if api_proto is not None and api_proto != service["protocol"]:
-                continue
-            if priority >= 100:
-                if service["priority"] == priority:
-                    no_results = False
-            elif service["priority"] < 100:
-                no_results = False
-        if no_results:
-            self._updateServices(srv_type)
+        valid_services = self._getValidServices(srv_type, priority, api_ver, api_proto)
 
-        # Re-check if there are any and return "" if not.
+        if len(valid_services) == 0:
+            self.updateServices(srv_type)
+            valid_services = self._getValidServices(srv_type, priority, api_ver, api_proto)
+
+            if len(valid_services) == 0:
+                raise NoService
+            else:
+                raise EndOfServiceList
+
+        # Randomise selection. Delete entry from the cached list of services and return it
+        random.seed()
+        index = random.randint(0, len(valid_services)-1)
+        service = valid_services[index]
+        href = self._createHref(service)
+        try:
+            self.services[srv_type].remove(service)
+        except Exception:
```

Comment:
When does this raise an exception?

Reply:
Don't think it can any more so will remove. 
Think I had a point where lists were out of sync